### ISSUE TITLE
About the design of neural network controller

### ISSUE THREAD

**Issue by Yngjx** [StanfordASL/neural-network-lyapunov#444]
Hi, I have read your code, but I still have some doubts. In these practical examples of inverted pendulum, I think the way you design the neural network controller is as follows:
1. first design the simple LQR controller;
2. then calculate the optimal controller gain matrix by minimizing the state error and the quadratic cost function of the control action;
3. finally use a neural network to approximate this optimal LQR controller.
I am not sure if my understanding is correct?
Thank you for reading and I look forward to your response!

**Comment by hongkai-dai** [comment]
Hi @Yngjx ,

We search two quantities, the controller `u = π(x)` and the Lyapunov function `V(x)`. Both the controller and the Lyapunov function contain neural networks, and require initialization. Here is how we do it.
1. First linearize the dynamics around the equilibrium, and find the LQR controller, with gain `u = K*x`, and a quadratic cost-to-go function `J(x) = x' * Q * x`.
2. Approximate the LQR controller `u = K*x` and the quadratic cost-to-go function with the neural network controller and Lyapunov functions through regression. Use this approximation as the initial guess.
3. Apply our algorithm to search for the neural network controller and Lyapunov functions, starting from the initial guess in step 2.

For more details, you can refer to our [paper](https://arxiv.org/pdf/2109.14152.pdf) . Let me know if you find it unclear.

Best,
Hongkai

**Comment by Yngjx** [comment]
Hi @hongkai-dai  ,
Thanks for your reply!

Regarding the third point you brought up, how exactly do you go about searching for neural network controllers?
I read your literature and the literature uses MIP, is this the method you use to search for neural network controllers? But this part emphasizes more on satisfying the lyapunov condition.

I am a bit confused.

Best，
Yngjx

**Comment by hongkai-dai** [comment]
We solve an MIP to **certify** the Lyapunov condition, not to search for neural network controller/Lyapunov functions. Instead, we search for neural network controller/Lyapunov functions through gradient based optimization.

To elaborate on this, we solve an MIP to find the counter examples that violate the Lyapunov condition for a given neural network controller and Lyapunov function. Mathematically, we define J(θ, π) as
```
J(θ, π)  = max_xₜ  V_θ(f(xₜ, π(xₜ)) − (1−κ)V_θ(xₜ)
```
Namely the Lyapunov function is parameterized by θ, and the controller is `u = π(x)`. J(θ, π) is the maximal violation of the Lyapunov condition V(xₜ₊₁) − V(xₜ) ≤ −κ V(xₜ) for a given controller and Lyapunov function candidate. We can compute this maximal violation through an MIP.

We then want to search for the controller and the Lyapunov function through the gradient-based minimization
```
min_{θ, π} J(θ, π)
```
where we have two algorithms to do this gradient based minimization:
1. We can minimize this cost on a training set of states, where the training set contains the counter example states found through MIP.
2. We can compute the gradient of `J` w.r.t `θ, π` through the KKT condition of the inner maximizing problem.

For more details on these two algorithms, please refer to section IV.B of the [paper](https://arxiv.org/pdf/2109.14152.pdf).

**Comment by Yngjx** [comment]
Thank you for your brilliant answer!

I understand the whole process of this design, you are really a serious and responsible researcher.
I will read your paper more carefully in my follow-up work.

I wish you a happy life!

**Comment by hongkai-dai** [comment]
Thanks @Yngjx , good luck with your research, and let me know if you have questions about our work.

I will mark this issue as closed.

**Comment by Yngjx** [comment]
OK, I'll ask you for advice if I have any follow-up questions.
